L1Trigger/TrackerTFP/src/DataFormats.cc

```diff
@@ -0,0 +1,707 @@
+  void DataFormats::fillDataFormats() {
+    if constexpr(config_[+v][+p] == p) {
+      dataFormats_.emplace_back(Format<v, p>(iConfig_, setup_));
+      fillFormats<v, p>();
+    }
+    if constexpr(++p != Process::end)
+      fillDataFormats<v, ++p>();
+    else if constexpr(++v != Variable::end)
+      fillDataFormats<++v>();
+  }
+
+  template<Variable v, Process p, Process it = Process::begin>
+  void DataFormats::fillFormats() {
+    if (config_[+v][+it] == p) {
+      formats_[+v][+it] = &dataFormats_.back();
+    }
+    if constexpr(++it != Process::end)
+      fillFormats<v, p, ++it>();
+  }
+
+  template<typename ...Ts>
+  void DataFormats::convertStub(const TTDTC::BV& bv, tuple<Ts...>& data, Process p) const {
+    TTBV ttBV(bv);
+    extractStub(ttBV, data, p);
+  }
+
+  template<int it = 0, typename ...Ts>
+  void DataFormats::extractStub(TTBV& ttBV, std::tuple<Ts...>& data, Process p) const {
+    Variable v = *next(stubs_[+p].begin(), sizeof...(Ts) - 1 - it);
+    formats_[+v][+p]->extract(ttBV, get<sizeof...(Ts) - 1 - it>(data));
+    if constexpr(it + 1 != sizeof...(Ts))
+      extractStub<it + 1>(ttBV, data, p);
+  }
+
+  template<typename... Ts>
+  void DataFormats::convertStub(const std::tuple<Ts...>& data, TTDTC::BV& bv, Process p) const {
+    TTBV ttBV(1, numUnusedBitsStubs_[+p]);
+    attachStub(data, ttBV, p);
+    bv = ttBV.bs();
+  }
+
+  template<int it = 0, typename... Ts>
+  void DataFormats::attachStub(const tuple<Ts...>& data, TTBV& ttBV, Process p) const {
+    Variable v = *next(stubs_[+p].begin(), it);
+    formats_[+v][+p]->attach(get<it>(data), ttBV);
+    if constexpr(it + 1 != sizeof...(Ts))
+      attachStub<it + 1>(data, ttBV, p);
+  }
+
+  template<typename ...Ts>
+  void DataFormats::convertTrack(const TTDTC::BV& bv, tuple<Ts...>& data, Process p) const {
+    TTBV ttBV(bv);
+    extractTrack(ttBV, data, p);
+  }
+
+  template<int it = 0, typename ...Ts>
+  void DataFormats::extractTrack(TTBV& ttBV, std::tuple<Ts...>& data, Process p) const {
+    Variable v = *next(tracks_[+p].begin(), sizeof...(Ts) - 1 - it);
+    formats_[+v][+p]->extract(ttBV, get<sizeof...(Ts) - 1 - it>(data));
+    if constexpr(it + 1 != sizeof...(Ts))
+      extractTrack<it + 1>(ttBV, data, p);
+  }
+
+  template<typename... Ts>
+  void DataFormats::convertTrack(const std::tuple<Ts...>& data, TTDTC::BV& bv, Process p) const {
+    TTBV ttBV(1, numUnusedBitsTracks_[+p]);
+    attachTrack(data, ttBV, p);
+    bv = ttBV.bs();
+  }
+
+  template<int it = 0, typename... Ts>
+  void DataFormats::attachTrack(const tuple<Ts...>& data, TTBV& ttBV, Process p) const {
+    Variable v = *next(tracks_[+p].begin(), it);
+    formats_[+v][+p]->attach(get<it>(data), ttBV);
+    if constexpr(it + 1 != sizeof...(Ts))
+      attachTrack<it + 1>(data, ttBV, p);
+  }
+
+  template<typename ...Ts>
+  Stub<Ts...>::Stub(const TTDTC::Frame& frame, const DataFormats* dataFormats, Process p) :
```

Comment:
The constructors of the Stub class use extremely dense C++. Please add a comment to each one explaining what it is used for, and what it is doing.

Reply:
done